cpp/src/arrow/array/data.h

```diff
@@ -230,7 +234,19 @@ struct ARROW_EXPORT ArrayData {
     return null_count.load() != length;
   }
 
-  // Access a buffer's data as a typed C pointer
+  /// \brief Access a buffer's data as a typed C pointer
+  ///
+  /// \param i the buffer index
+  /// \param absolute_offset the offset into the buffer
+  ///
+  /// If `absolute_offset` is non-zero, the type `T` must match the
+  /// layout of buffer number `i` for the array's data type; otherwise
+  /// offset computation would be incorrect.
+  ///
+  /// If the given buffer is bit-packed (such as a validity bitmap, or
+  /// the data buffer of a boolean array), then `absolute_offset` must be
+  /// zero for correct results, and any bit offset must be applied manually
+  /// by the caller.
```

Comment:
Perhaps instead we should avoid use of GetValues* for bit packed buffers altogether
```suggestion
  /// This method is not suitable for accessing the data of a buffer which
  /// is bit-packed (such as a validity bitmap, or the data buffer of a boolean array).
  /// To access bits, use `buffers[i]->data()` and apply the bit offset manually.
```

Reply:
Well, `GetValues` protects against missing buffers and non-CPU buffers (the former is important as the validity bitmap is optional).